src/python_testing/TC_TSTAT_4_2.py

```diff
@@ -461,7 +465,41 @@ async def test_TC_TSTAT_4_2(self):
             # Roll back
             await self.send_atomic_request_rollback_command()
 
-        # TODO: Add tests for the total number of Presets exceeds the NumberOfPresets supported. Also Add tests for adding presets with preset scenario not present in PresetTypes.
+        self.step("17")
+        if self.pics_guard(self.check_pics("TSTAT.S.F08") and self.check_pics("TSTAT.S.A0050") and self.check_pics("TSTAT.S.Cfe.Rsp")):
+            test_presets = new_presets_with_handle.copy()
+            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=9,
+                                                             name="Wake", coolingSetpoint=2500, heatingSetpoint=1700, builtIn=False))
+
+            # Send the AtomicRequest begin command
+            await self.send_atomic_request_begin_command()
+
+            await self.write_presets(endpoint=endpoint, presets=test_presets, expected_status=Status.ConstraintError)
+
+            # Clear state for next test.
+            await self.send_atomic_request_rollback_command()
+
+        self.step("18")
+        if self.pics_guard(self.check_pics("TSTAT.S.F08") and self.check_pics("TSTAT.S.A0050") and self.check_pics("TSTAT.S.Cfe.Rsp")):
+
+            # Read the active preset handle attribute and verify it was updated to preset handle
+            presetTypes = await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=cluster.Attributes.PresetTypes)
+            logger.info(f"Rx'd PresetTypes: {presetTypes}")
+
+            test_presets = copy.deepcopy(new_presets_with_handle)
+            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=cluster.Enums.PresetScenarioEnum.kWake,
+                                                             name="Wake", coolingSetpoint=2500, heatingSetpoint=1700, builtIn=False))
+            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=cluster.Enums.PresetScenarioEnum.kVacation,
+                                                             name="Vacation", coolingSetpoint=2400, heatingSetpoint=1800, builtIn=False))
+            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=cluster.Enums.PresetScenarioEnum.kGoingToSleep,
+                                                             name="GoingToSleep", coolingSetpoint=2300, heatingSetpoint=1900, builtIn=False))
+            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=cluster.Enums.PresetScenarioEnum.kUserDefined,
+                                                             name="UserDefined", coolingSetpoint=2300, heatingSetpoint=1900, builtIn=False))
+
+            # Send the AtomicRequest begin command
+            await self.send_atomic_request_begin_command()
+
+            await self.write_presets(endpoint=endpoint, presets=test_presets, expected_status=Status.ResourceExhausted)
```

Comment:
Should we add a rollback here as well?

Reply:
maybe as a separate PR. i will add a new issue and tests

Reply:
done.